pyro/distributions/transforms/spline.py

```diff
@@ -33,15 +37,28 @@ def _select_bins(x, idx):
     Performs gather to select the bin in the correct way on batched inputs
     """
     idx = idx.clamp(min=0, max=x.size(-1) - 1)
-    extra_dims = len(idx.shape) - len(x.shape)
-    expanded_x = x.expand(idx.shape[:extra_dims] + (-1,) * len(x.shape))
-    return expanded_x.gather(-1, idx).squeeze(-1)
+
+    """
+    idx ~ (context_batch_dims, input_dim, 1)
+    x ~ (batch_dims, input_dim, count_bins)
+    """
+
+    # Broadcast dimensions of idx over x
+    if len(idx.shape) >= len(x.shape):
+        x = x.reshape((1,) * (len(idx.shape) - len(x.shape)) + x.shape)
+        x = x.expand(idx.shape[:-2] + (-1,) * 2)
+    # Broadcast dimensions of x over idx
+    elif len(idx.shape) <= len(x.shape):
+        idx = idx.reshape((1,) * (len(x.shape) - len(idx.shape)) + idx.shape)
+        idx = idx.expand(x.shape[:-2] + (-1,) * 2)
```

Comment:
These conditions overlap and look dubious. Why do you use `<=` and `>=` rather than strict inequalities? Why is `.expand()` logic gated by `.dim()` conditions rather than `.shape` conditions?

Reply:
I have to look into this again... What I'm trying to take into account is that the context variable may have batch dimensions, which then effect the batch dimensions of the variable `idx`. We have to expand `idx` so that its batch dimensions match up with that of `x` before performing `gather`.

What I mean by batch dimensions on `context` is that we can have a single context variable that is produces parameters that are shared by all elements of the input variable batch, or we can have a batch of context variables so that each input variable in the batch has its own parameters

Reply:
See the latest version for a resolution of this issue